cmd/kubeadm/app/cmd/phases/join/controlplanejoin.go

```diff
@@ -0,0 +1,253 @@
+	// the existing etcd members before being created.
+	// This operation must be executed after kubelet is already started in order to minimize the time
+	// between the new etcd member is announced and the start of the static pod running the new etcd member, because during
+	// this time frame etcd gets temporary not available (only when moving from 1 to 2 members in the etcd cluster).
+	// From https://coreos.com/etcd/docs/latest/v2/runtime-configuration.html
+	// "If you add a new member to a 1-node cluster, the cluster cannot make progress before the new member starts
+	// because it needs two members as majority to agree on the consensus. You will only see this behavior between the time
+	// etcdctl member add informs the cluster about the new member and the new member successfully establishing a connection to the 	// existing one."
+	if err := etcdphase.CreateStackedEtcdStaticPodManifestFile(client, kubeadmconstants.GetStaticPodDirectory(), cfg.NodeRegistration.Name, &cfg.ClusterConfiguration, &cfg.LocalAPIEndpoint); err != nil {
+		return errors.Wrap(err, "error creating local etcd static pod manifest file")
+	}
+
+	return nil
+}
+
+func runUploadConfigPhase(c workflow.RunData) error {
+	data, ok := c.(controlPlaneJoinData)
+	if !ok {
+		return errors.New("control-plane-join phase invoked with an invalid data struct")
+	}
+
+	kubeConfigFile := filepath.Join(kubeadmconstants.KubernetesDir, kubeadmconstants.AdminKubeConfigFileName)
+
+	client, err := data.ClientSetFromFile(kubeConfigFile)
+	if err != nil {
+		return errors.Wrap(err, "couldn't create Kubernetes client")
+	}
+	cfg, err := data.InitCfg()
+	if err != nil {
+		return err
+	}
+
+	if err := uploadconfigphase.UploadConfiguration(cfg, client); err != nil {
+		return errors.Wrap(err, "error uploading configuration")
+	}
+
+	return nil
+}
+
+func runMarkControlPlanePhase(c workflow.RunData) error {
+	data, ok := c.(controlPlaneJoinData)
+	if !ok {
+		return errors.New("control-plane-join phase invoked with an invalid data struct")
+	}
+
+	kubeConfigFile := filepath.Join(kubeadmconstants.KubernetesDir, kubeadmconstants.AdminKubeConfigFileName)
+
+	client, err := data.ClientSetFromFile(kubeConfigFile)
+	if err != nil {
+		return errors.Wrap(err, "couldn't create Kubernetes client")
+	}
+	cfg, err := data.InitCfg()
+	if err != nil {
+		return err
+	}
+
+	if err := markcontrolplanephase.MarkControlPlane(client, cfg.NodeRegistration.Name, cfg.NodeRegistration.Taints); err != nil {
+		return errors.Wrap(err, "error applying control-plane label and taints")
+	}
+
+	return nil
+}
+
+func runControlPlaneJoinComponentSubphase(component string) func(c workflow.RunData) error {
+	return func(c workflow.RunData) error {
+		data, ok := c.(controlPlaneJoinData)
+		if !ok {
+			return errors.New("control-plane-join phase invoked with an invalid data struct")
+		}
+
+		// Skip if this is not a control plane
+		if data.Cfg().ControlPlane == nil {
```

Comment:
Now this concerns me kind of indirectly. If I am not mistaken, `data.Cfg().ControlPlane` is going to be different from `nil` only if `--experimental-control-plane` is specified at the command line or it's filled in the config file. Hence, if the is no config file, the UX is going to be a bit odd, because we cannot execute the control-plane-join phase without the above flag at the command line.

The mandatory command line without config file would look something like this:
```
kubeadm join --experimental-control-plane phase control-plane-join ...
```

I believe, that we need to fix this.

@fabriziopandini @neolit123 WDYT?

Reply:
Yes, this is the case, but as @fabriziopandini pointed out this is already taken care by the control plane prepare phase (and needs the flag if no config file is provided).